### PR TITLE
Upgrade SQLAlchemy from 1.3.7 to 1.3.17; Fix #4055

### LINKED ISSUE #4055: Upgrade SQLAlchemy version
Motivation
----------
Upgrade to the newer version of SQLAlchemy. This will probably require some changes in the code.

### PR DESCRIPTION
**Pull request by mlassnig**
1.3.8 had a conceptual change in the ENUM handling on PostgreSQL.

### CODE REVIEW

**Comment by mlassnig** [comment]
Please have a close look y'all :-) I'm pretty sure might also affects some daemons, but these should be fixed in a second step.

File: lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py [hunk 1]
```diff
@@ -1,3 +1,21 @@
...
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Martin Barisits <martin.barisits@cern.ch>, 2019
+# - Mario Lassnig <mario.lassnig@cern.ch>, 2019-2020
 # Copyright 2019 CERN for the benefit of the ATLAS collaboration.
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py]: double copyright

File: tools/run_tests_docker.sh [hunk 2]
```diff
@@ -149,8 +150,13 @@ if test ${special}; then
...
     pytest -v --full-trace lib/rucio/tests/test_dirac.py
 else
-    echo 'Running tests'
-    pytest -v --full-trace
+    if test ${notrace}; then
+	echo 'Running tests'
+	pytest -v --tb=short
+    else
+	echo 'Running tests with full tracing'
+	pytest -v --full-trace
```
Comment (bziemons) [inline tools/run_tests_docker.sh]: I have already removed the "full-trace" for regular test runs. I realized the normal output also contains the trace without code output for each stack frame. And I agree it was too much output to have a good overview.

File: tools/run_tests_docker.sh [hunk 3]
```diff
@@ -149,8 +150,13 @@ if test ${special}; then
     echo 'Using the special config and only running test_dirac'
     pytest -v --full-trace lib/rucio/tests/test_dirac.py
 else
-    echo 'Running tests'
-    pytest -v --full-trace
+    if test ${notrace}; then
```
Comment (bziemons) [inline tools/run_tests_docker.sh]: this is a good idea nonetheless (see comment below).

File: lib/rucio/db/sqla/migrate_repo/versions/1a29d6a9504c_add_didtype_chck_to_requests.py [hunk 4]
```diff
@@ -1,3 +1,22 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Vincent Garonne <vincent.garonne@cern.ch>, 2015-2017
+# - Martin Barisits <martin.barisits@cern.ch>, 2016
+# - Mario Lassnig <mario.lassnig@cern.ch>, 2019-2020
 # Copyright 2013-2019 CERN for the benefit of the ATLAS collaboration.
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/1a29d6a9504c_add_didtype_chck_to_requests.py]: double copyright

File: lib/rucio/db/sqla/migrate_repo/versions/01eaf73ab656_add_new_rule_notification_state_progress.py [hunk 5]
```diff
@@ -1,3 +1,22 @@
...
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+#
+# Authors:
+# - Martin Barisits <martin.barisits@cern.ch>, 2019
+# - Mario Lassnig <mario.lassnig@cern.ch>, 2019-2020
+# - Robert Illingworth <illingwo@fnal.gov>, 2019
 # Copyright 2019 CERN for the benefit of the ATLAS collaboration.
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/01eaf73ab656_add_new_rule_notification_state_progress.py]: double copyright

**Review by bziemons — DISMISSED** [review]

File: tools/run_tests_docker.sh (hunk 3 above)
Comment (bziemons) [reply tools/run_tests_docker.sh]: i.e. make this option remove the `-vvvrxs` from the pytest call.

**Comment by mlassnig** [comment]
ok, that was nasty. sqlalchemy 1.3.18 introduced a bug that generates bad sql syntax for nowait on mysql5 databases. if we want to continue to support mysql5 we can't upgrade beyond 1.3.17 (or until this is fixed in sqlalchemy). i will clean this patch up tomorrow, then it's ready for merging. @bari12 @bziemons

File: etc/pip-requires [hunk 6]
```diff
@@ -1,9 +1,11 @@
...
-SQLAlchemy==1.3.7                                 # DB backend
-alembic==1.4.1                                    # Lightweight database migration tool for SQLAlchemy
-web.py==0.62                                      # Python web framework
-python3-memcached==1.51                           # Quick and small memcached client for Python3
+SQLAlchemy==1.3.17                                # DB backend
+alembic==1.4.3                                    # Lightweight database migration tool for SQLAlchemy
+web.py==0.39; python_version <= '2.7'             # Python web framework for Python2
+web.py==0.40; python_version > '2.7'              # Python web framework for Python3
+python-memcached==1.59; python_version <= '2.7'   # Quick and small memcached client for Python2
+python3-memcached==1.51; python_version > '2.7'   # Quick and small memcached client for Python3
```
Comment (bziemons) [inline etc/pip-requires]: `python3-memcached` had its last release in 2013, while `python-memcached` is newer -- removing `python3-memcached` and only using `python-memcached` seems to be the best option

File: lib/rucio/client/baseclient.py [hunk 7]
```diff
@@ -80,11 +85,17 @@
...
+        try:
+            imp.find_module(extra_module)
+            EXTRA_MODULES[extra_module] = True
+        except ImportError:
+            EXTRA_MODULES[extra_module] = False
+    else:
+        if importlib.util.find_spec(extra_module):
+            EXTRA_MODULES[extra_module] = True
+        else:
+            EXTRA_MODULES[extra_module] = False
```
Comment (bziemons) [inline lib/rucio/client/baseclient.py:98]: can this be moved to a util function? I have seen similar code.

(can also be solved in a follow-up)

File: lib/rucio/common/utils.py [hunk 8]
```diff
@@ -92,18 +101,18 @@
...
+        try:
+            imp.find_module(extra_module)
+            EXTRA_MODULES[extra_module] = True
+        except ImportError:
+            EXTRA_MODULES[extra_module] = False
+    else:
+        if importlib.util.find_spec(extra_module):
+            EXTRA_MODULES[extra_module] = True
+        else:
+            EXTRA_MODULES[extra_module] = False
```
Comment (bziemons) [inline lib/rucio/common/utils.py:115]: see comment above.

File: lib/rucio/core/message.py [hunk 9]
```diff
@@ -107,18 +106,25 @@ def retrieve_messages(bulk=1000, thread=None, total_threads=None, event_type=Non
         # Step 1:
         # MySQL does not support limits in nested queries, limit on the outer query instead.
         # This is not as performant, but the best we can get from MySQL.
+        # FIXME: SQLAlchemy generates wrong nowait MySQL8 statement for MySQL5
+        #        Remove once this is resolved in SQLAlchemy
```
Comment (bziemons) [inline lib/rucio/core/message.py:110]: I thought this was solved by downgrading SQLAlchemy to 1.3.17?

File: lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py [hunk 10]
```diff
@@ -35,11 +35,14 @@ def upgrade():
...
     schema = context.get_context().version_table_schema + '.' if context.get_context().version_table_schema else ''
 
-    if context.get_context().dialect.name in ['oracle', 'postgresql']:
+    if context.get_context().dialect.name == 'oracle':
         drop_constraint('REPLICAS_STATE_CHK', 'replicas', type_='check')
         create_check_constraint(constraint_name='REPLICAS_STATE_CHK', table_name='replicas',
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'S', 'T')")
 
+    elif context.get_context().dialect.name == 'postgresql':
+        pass  # too complicated on PostgreSQL -- leave the ENUM as it is
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py]: Didn't you already do this above?

https://github.com/rucio/rucio/blob/5cc5020284ea490386fb1f97f53f4f0674abdfec/lib/rucio/db/sqla/migrate_repo/versions/01eaf73ab656_add_new_rule_notification_state_progress.py#L45-L48

File: lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py [hunk 11]
```diff
@@ -63,9 +66,7 @@ def downgrade():
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'S')")
 
     elif context.get_context().dialect.name == 'postgresql':
-        op.execute('ALTER TABLE ' + schema + 'replicas DROP CONSTRAINT IF EXISTS "REPLICAS_STATE_CHK", ALTER COLUMN state TYPE CHAR')  # pylint: disable=no-member
-        create_check_constraint(constraint_name='REPLICAS_STATE_CHK', table_name='replicas',
-                                condition="state in ('A', 'U', 'C', 'B', 'D', 'S')")
+        pass  # too complicated on PostgreSQL -- leave the ENUM as it is
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py]: see comment above.

File: lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py [hunk 12]
```diff
@@ -34,14 +35,17 @@ def upgrade():
...
+    if context.get_context().dialect.name == 'oracle':
         drop_constraint('REPLICAS_STATE_CHK', 'replicas', type_='check')
         create_check_constraint(constraint_name='REPLICAS_STATE_CHK', table_name='replicas',
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'T')")
         drop_constraint('COLLECTION_REPLICAS_STATE_CHK', 'collection_replicas', type_='check')
         create_check_constraint(constraint_name='COLLECTION_REPLICAS_STATE_CHK', table_name='collection_replicas',
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'T')")
 
+    elif context.get_context().dialect.name == 'postgresql':
+        pass  # too complicated in PostgreSQL -- leave the ENUM as it is
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py]: see comment above.

File: lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py [hunk 13]
```diff
@@ -62,26 +65,25 @@ def downgrade():
...
+    if context.get_context().dialect.name == 'oracle':
         drop_constraint('REPLICAS_STATE_CHK', 'replicas', type_='check')
         create_check_constraint(constraint_name='REPLICAS_STATE_CHK', table_name='replicas',
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'S', 'T')")
         drop_constraint('COLLECTION_REPLICAS_STATE_CHK', 'collection_replicas', type_='check')
         create_check_constraint(constraint_name='COLLECTION_REPLICAS_STATE_CHK', table_name='collection_replicas',
                                 condition="state in ('A', 'U', 'C', 'B', 'D', 'S', 'T')")
 
+    elif context.get_context().dialect.name == 'postgresql':
+        pass  # too complicated in PostgreSQL -- leave the ENUM as it is
```
Comment (bziemons) [inline lib/rucio/db/sqla/migrate_repo/versions/b7d287de34fd_removal_of_replicastate_source.py]: see comment above.

File: etc/pip-requires (hunk 6 above)
Comment (mlassnig) [reply etc/pip-requires]: o.O this comes from martin's patch from the upstream/master rebase. not touching that ;-)

File: lib/rucio/client/baseclient.py (hunk 7 above)
Comment (mlassnig) [reply lib/rucio/client/baseclient.py:98]: follow-up sounds good. this was just one of the random improvements i did while waiting for the github actions to finish.

File: lib/rucio/core/message.py (hunk 9 above)
Comment (mlassnig) [reply lib/rucio/core/message.py:110]: correct, will take this out.

File: lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py (hunk 10 above)
Comment (mlassnig) [reply lib/rucio/db/sqla/migrate_repo/versions/7ec22226cdbf_new_replica_state_for_temporary_.py]: will check

File: etc/pip-requires (hunk 6 above)
Comment (bziemons) [reply etc/pip-requires]: then you should revert your changes to these lines, no?

Comment (mlassnig) [reply etc/pip-requires]: not sure why they show up at all, i didn't put them. all the other changes from the rebase don't show up either (correctly). hm.

**Comment by bziemons** [comment]
I realized in #4079 when I rebased on this that postgres and mysql8 (but not oracle) are in different states regarding the checks (check does not exist in mysql8 but in oracle/postgres it is present). Isn't this an issue?

**Comment by mlassnig** [comment]
> I realized in #4079 when I rebased on this that postgres and mysql8 (but not oracle) are in different states regarding the checks (check does not exist in mysql8 but in oracle/postgres it is present). Isn't this an issue?

which ones exactly?

**Comment by bziemons** [comment]
> > I realized in #4079 when I rebased on this that postgres and mysql8 (but not oracle) are in different states regarding the checks (check does not exist in mysql8 but in oracle/postgres it is present). Isn't this an issue?
> 
> which ones exactly?

I am talking about the `REQUESTS_STATE_CHK` but thats probably not the exception then

File: etc/pip-requires (hunk 6 above)
Comment (bari12) [reply etc/pip-requires]: I will address this in a separate PR.

File: lib/rucio/core/message.py (hunk 9 above)
Comment (bziemons) [reply lib/rucio/core/message.py:110]: still in there? anything wrong?

Comment (mlassnig) [reply lib/rucio/core/message.py:110]: yeah, seems like this particular statement still requires this. not exactly sure why. the other NOWAIT errors all across the core are gone though.

**Review by bziemons — APPROVED** [review]

**Review by bari12 — APPROVED** [review]

**Pull request merged**